metricflow/dataflow/builder/dataflow_plan_builder.py

```diff
@@ -836,7 +830,8 @@ def _find_dataflow_recipe(
         # MetricGroupBy source nodes could be extremely large (and potentially slow).
         candidate_nodes_for_right_side_of_join += [
             self._build_query_output_node(
-                query_spec=self._query_spec_for_source_node(group_by_metric_spec), for_group_by_source_node=True
+                query_spec=self._query_parser.build_query_spec_for_group_by_metric_source_node(group_by_metric_spec),
```

Comment:
Ok, I did a fair bit of digging and in order to avoid this query parser -> dataflow plan builder -> query parser -> dataflow plan builder loop we'd have to re-order most of the meaningful operations between engine initialization and here, so I don't think there's any other reasonable way to do this.

Another options we might consider is moving this into the SourceNodeBuilder and making the SourceNodeBuilder a property of the DataflowPlanBuilder. Then we could access these as a set of "metric source nodes" and avoid direct access to the query parser here.

Reply:
I have no problem doing that - but curious why that seems better to you?

Reply:
i.e., what is so bad about wiring the query parser through here?

Reply:
The downside of directly importing the query parser is it opens the door to weird cycles where we could, at any point in dataflow plan construction, decide to parse another query and build another dataflow plan. Super powerful, hard to reason about, so I'd rather avoid making the query parser available here until we really have no other choice. If everything is behind the source node builder we make it more obvious that these special processing directives are scope-limited to source nodes, rather than something we can apply to arbitrary mid-DAG nodes.

There's also an upside to moving this call to the SourceNodeBuilder - these GroupByMetric objects are a special type of source node in the DAG. Even though they'd be typed differently from the existing source nodes, bundling their accessor in the provider we created for source nodes feels more natural to me than doing an inlined query parser call.

So on balance I have a preference for moving this out if that doesn't cause annoying dependency cycles somewhere else.